pytorch/pytorch
# [Don't Review Yet] Make wait_tensor as an effect ops

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132156
* __->__ #132155
* #131351

Right now `wait_tensor()` may be DCEd if its result is not used. This will cause desync issues. This PR makes `wait_tensor` as an effect ops.

Command to verify:
```
pytest test/distributed/_tensor/test_attention.py  -k test_ring_attention_sdpa_is_causal_True_compiled_True_backend1
```

What need to be done:
1. Add `torch.ops._c10d_functional.wait_tensor.default` to SIDE_EFFECTS.
2. Ensure Inductor can correctly lower the op.
3. Change the existing PTD passes (DDP, async TP, and more?) to handle the with_effect op.
